engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala

```diff
@@ -0,0 +1,340 @@
+  private def processCallArguments(
+    arguments: List[CallArgument],
+    graph: Graph
+  ): List[CallArgument] = {
+    arguments.map { case arg @ CallArgument.Specified(name, value, _, _, _) =>
+      maybeAttachFramePointer(arg, graph)
+        .copy(
+          name  = name.map(maybeAttachFramePointer(_, graph)),
+          value = processExpression(value, graph)
+        )
+    }
+  }
+
+  /** Attaches [[FramePointerMeta]] metadata to the given `ir` if there is an
+    * appropriate [[Info.Occurrence]] already attached to it.
+    * @param ir IR to attach the frame pointer metadata to.
+    * @param graph Alias analysis graph
+    * @tparam T Type of IR.
+    * @return Copy of `ir` with attached metadata, or just the `ir` if nothing
+    *         was attached.
+    */
+  private def maybeAttachFramePointer[T <: IR](
+    ir: T,
+    graph: Graph
+  ): T = {
+    getAliasAnalysisMeta(ir) match {
+      case Some(Info.Occurrence(_, id)) =>
+        graph.scopeFor(id) match {
+          case Some(scope) =>
+            graph.getOccurrence(id) match {
+              case Some(use: Graph.Occurrence.Use) =>
+                // Use is allowed to read a variable from some parent scope
+                graph.defLinkFor(use.id) match {
+                  case Some(defLink) =>
+                    val defId = defLink.target
+                    val defOcc = graph
+                      .getOccurrence(defId)
+                      .get
+                      .asInstanceOf[Graph.Occurrence.Def]
+                    val defScope    = graph.scopeFor(defId).get
+                    val parentLevel = getScopeDistance(defScope, scope)
+                    val frameSlotIdx =
+                      getFrameSlotIdxInScope(graph, defScope, defOcc)
+                    ir.updateMetadata(
+                      new MetadataPair(
+                        this,
+                        new FramePointerMeta(
+                          new FramePointer(parentLevel, frameSlotIdx)
+                        )
+                      )
+                    )
+                  case None =>
+                    // It is possible that there is no Def for this Use. It can, for example, be
+                    // Use for some global symbol. In `IrToTruffle`, an UnresolvedSymbol will be
+                    // generated for it.
+                    // We will not attach any metadata in this case.
+                    ir
+                }
+              case Some(defn: Graph.Occurrence.Def) =>
+                // The definition cannot write to parent's frame slots.
+                val parentLevel  = 0
+                val frameSlotIdx = getFrameSlotIdxInScope(graph, scope, defn)
+                ir.updateMetadata(
+                  new MetadataPair(
+                    this,
+                    new FramePointerMeta(
+                      new FramePointer(parentLevel, frameSlotIdx)
+                    )
+                  )
+                )
+              case _ => ir
```

Comment:
We seem to do a lot of `.copy` in other methods, but here, at the core method, we just _update the IR_ and return the same `ir` reference back. This makes me believe all the copying is unnecessary.

Reply:
All copying removed in 61821bf3932f72bac1355f2aed3c05af6972d02b